File: leetcode/338-CountingBits.py
class Solution:
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        result = [-1]*(num + 1)
        result[0] = 0
        for i in range(num + 1):
            if i == 0:
                result[i] = 0
            else:
                log = math.log2(i)
                if log.is_integer():
                    result[i] = 1
                else:
                    result[i] = result[i - 2**int(log)] + 1
        return result

File: leetcode/338-CountingBits.java
class Solution {
    public int[] countBits(int num) {
        if(num == 0){
            return new int[]{0};
        }
        int[] res = new int[num + 1];
        res[0] = 0;
        int prePower = 0;
        for(int i = 1;i < res.length; i++){
            if((i & (i - 1)) == 0){
                // if is 2's pow
                res[i] = 1;
                prePower = i;
                
            }
            else{
                res[i] = res[i - prePower] + 1;
            }
        }
        return res;
    }
}
